src/qdwh.cc

```diff
@@ -0,0 +1,410 @@
+// Copyright (c) 2017-2022, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include "slate/slate.hh"
+#include "auxiliary/Debug.hh"
+#include "slate/Matrix.hh"
+#include "internal/internal.hh"
+
+namespace slate {
+
+namespace impl {
+
+//------------------------------------------------------------------------------
+/// Distributed parallel polar decomposition based on QDWH algorithm.
+/// Generic implementation for any target.
+/// @ingroup qdwh_specialization
+///
+template <Target target, typename scalar_t>
+void qdwh(
+          Matrix<scalar_t>& A,
+          Matrix<scalar_t>& H,
+          int& itqr, int& itpo,
+          Options const& opts)
+{
+    // todo:
+    // optimizations:
+    // 1. reuse Q from first geqrf
+    // 2. avoid rounding m if m is not divisible by nb, because then geqrf/unmqr
+    // have extra zero rows
+    // in norm2est, when allocate vector mx1, need to allocate on row-processes only
+    using real_t = blas::real_type<scalar_t>;
+    using blas::real;
+
+    int64_t max_panel_threads  = std::max(omp_get_max_threads()/2, 1);
+
+    max_panel_threads = get_option<int64_t>( opts, Option::MaxPanelThreads, max_panel_threads );
+
+    int itconv, it;
+
+    real_t L2, sqd, dd, a1, a, b, c;
+    real_t Li, Liconv;
+    real_t conv = 100.;
+    real_t rzero = 0.0, rone = 1.0;
+
+    scalar_t zero = 0.0, one = 1.0;
+    scalar_t alpha, beta;
+
+    real_t normR;
+
+    real_t eps  = std::numeric_limits<real_t>::epsilon();
```

Comment:
Note this definition of epsilon is unit roundoff or ulp, whereas (Sca)LAPACK uses eps = 2*ulp, e.g. in `lamch`. Would that affect any of your formulas?

Reply:
In the matlab code of qdwh, eps is the unit roundoff, so I am using the same here.